crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -379,10 +379,40 @@ impl<'a> Checker<'a> {
         self.indexer.comment_ranges()
     }
 
-    /// Push a new [`Diagnostic`] to the collection in the [`Checker`]
-    pub(crate) fn report_diagnostic(&self, diagnostic: Diagnostic) {
-        let mut diagnostics = self.diagnostics.borrow_mut();
-        diagnostics.push(diagnostic);
+    /// Return a [`DiagnosticGuard`] for reporting a diagnostic.
+    ///
+    /// The guard derefs to a [`Diagnostic`], so it can be used to further modify the diagnostic
+    /// before it is added to the collection in the checker on `Drop`.
+    pub(crate) fn report_diagnostic<'chk, T: Violation>(
+        &'chk self,
+        kind: T,
+        range: TextRange,
+    ) -> DiagnosticGuard<'chk, 'a> {
+        DiagnosticGuard {
+            checker: self,
+            diagnostic: Some(Diagnostic::new(kind, range)),
+        }
+    }
+
+    /// Return a [`DiagnosticGuard`] for reporting a diagnostic if the corresponding rule is
+    /// enabled.
+    ///
+    /// Prefer [`Checker::report_diagnostic`] in general because the conversion from a `Diagnostic`
+    /// to a `Rule` is somewhat expensive.
+    pub(crate) fn checked_report_diagnostic<'chk, T: Violation>(
+        &'chk self,
+        kind: T,
+        range: TextRange,
+    ) -> Option<DiagnosticGuard<'chk, 'a>> {
+        let diagnostic = Diagnostic::new(kind, range);
+        if self.enabled(diagnostic.rule()) {
```

Comment:
It's unfortunate that this has to go through `diagnostic.rule` which requires a lookup. I'm leaning towards adding a `rule` method to `ViolationMetadata`. That should be easy enough to derive and does make sense to me. This will also allow us to resolve the noqa code in a future version.

That makes me wonder if `report_diagnostic` should always to the `self::enabled` call but I suspect that handling with the `Option` return type is too annoying in many cases that it isn't worth it.